### PR TITLE
Sentences capitalization improving

### PR DESCRIPTION
**Pull request by Merkost**
Moved capitalization logic in a separate function
Added inputTypeClassVariation for controlling whether keyboard is in password mode to turn off capitalization

### CODE REVIEW

File: app/src/main/kotlin/com/simplemobiletools/keyboard/helpers/Constants.kt [hunk 1]
```diff
@@ -35,24 +38,30 @@ enum class ShiftState {
             }
         }
 
-        fun getCapitalizationOnDelete(context: Context, text: CharSequence?): ShiftState {
-            if (!context.config.enableSentencesCapitalization) {
-                return OFF
+        fun shouldCapitalize(context: Context, text: String?): Boolean {
+            //To capitalize first letter in textField
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/helpers/Constants.kt:42]: what does this mean? Also if you want to write a comment, add a space after "//" and start with lowercase

**Comment by tibbi** [comment]
emails start with a capital letter too, even if they shouldnt. Test it properly please and capitalize casual sentences only

**Comment by Merkost** [comment]
> emails start with a capital letter too, even if they shouldnt. Test it properly please and capitalize casual sentences only

Of course it won't work with email inputType, I have excluded capitalization only for the password inputType

**Comment by tibbi** [comment]
alright, try thinking of some other usecases too or check the available input types, more exceptions might be needed once we roll out the app

**Comment by tibbi** [comment]
it is still capital in too many places, for example Facebook search, attendees field in Simple Calendar (CalDAV synced event). Although that attendee field has no extra flag or input type, it is lowercase in Gboard. Id like to have it lowercase too.

**Comment by Merkost** [comment]
> it is still capital in too many places, for example Facebook search, attendees field in Simple Calendar (CalDAV synced event). Although that attendee field has no extra flag or input type, it is lowercase in Gboard. Id like to have it lowercase too.

Yep, I'm still on it. Thanks for testing, I'll add these exceptions.
However, will it be easy to just add some types where to use capitalizing? Instead of exceptions inputTypes

**Comment by tibbi** [comment]
ye, maybe we should use whitelist instead of blacklist, should be more reliable

File: app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt [hunk 2]
```diff
@@ -64,23 +62,64 @@ class SimpleKeyboardIME : InputMethodService(), MyKeyboardView.OnKeyboardActionL
...
+                    if (currentInputConnection.getCursorCapsMode(editorInfo.inputType) != 0) {
+                        keyboard?.setShifted(ShiftState.ON_ONE_CHAR)
+                        keyboardView?.invalidateAllKeys()
+                        return
+                    }
                 }
             }
         }
+
+        // In other cases reset shift to OFF
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt:120]: use lowercase I

File: app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt [hunk 3]
```diff
@@ -89,26 +128,31 @@ class SimpleKeyboardIME : InputMethodService(), MyKeyboardView.OnKeyboardActionL
             return
         }
 
+//        this.keyboardView.setEditorInfo(EditorInfo)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt:131]: do not comment out code, just remove it

File: app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt [hunk 4]
```diff
@@ -89,26 +128,31 @@ class SimpleKeyboardIME : InputMethodService(), MyKeyboardView.OnKeyboardActionL
...
+                    if (keyboard?.mShiftState != ShiftState.ON_PERMANENT) {
+                        keyboard?.setShifted(ShiftState.getShiftStateForText(this, inputTypeClassVariation, text?.toString()))
+                        keyboardView?.invalidateAllKeys()
+                    }
+
                     inputConnection.sendKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL))
                     inputConnection.sendKeyEvent(KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_DEL))
                 } else {
+
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt:153]: do not create useless empty lines like this

File: app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt [hunk 5]
```diff
@@ -186,26 +225,21 @@ class SimpleKeyboardIME : InputMethodService(), MyKeyboardView.OnKeyboardActionL
...
     }
 
     override fun onActionUp() {
         if (switchToLetters) {
             // TODO: Change keyboardMode to enum class
             keyboardMode = KEYBOARD_LETTERS
-            val text = currentInputConnection?.getExtractedText(ExtractedTextRequest(), 0)?.text
-            val newShiftState = ShiftState.getShiftStateForText(this, text?.toString().orEmpty())
+
+            //Check if capitalization is needed after switching to letters layout
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt:240]: again, add space and capital letter

File: app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt [hunk 6]
```diff
@@ -96,19 +103,14 @@ class SimpleKeyboardIME : InputMethodService(), MyKeyboardView.OnKeyboardActionL
...
-                    val extractedText = inputConnection.getTextBeforeCursor(3, 0)?.dropLast(1)
-                    keyboard!!.setShifted(ShiftState.getCapitalizationOnDelete(context = this, text = extractedText))
-                }
-
                 val selectedText = inputConnection.getSelectedText(0)
                 if (TextUtils.isEmpty(selectedText)) {
                     inputConnection.sendKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL))
                     inputConnection.sendKeyEvent(KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_DEL))
                 } else {
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt:111]: avoid adding useless empty lines

File: app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt [hunk 7]
```diff
@@ -64,23 +62,32 @@ class SimpleKeyboardIME : InputMethodService(), MyKeyboardView.OnKeyboardActionL
...
+        val editorInfo = currentInputEditorInfo
+        if (config.enableSentencesCapitalization && editorInfo != null && editorInfo.inputType != InputType.TYPE_NULL) {
+            if (currentInputConnection.getCursorCapsMode(editorInfo.inputType) != 0) {
+                keyboard?.setShifted(ShiftState.ON_ONE_CHAR)
+                keyboardView?.invalidateAllKeys()
+                return
             }
         }
+
+        // in other cases reset shift to OFF
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/keyboard/services/SimpleKeyboardIME.kt:88]: avoid adding useles comments too please, it is self explanatory enough

**Comment by tibbi** [comment]
functionality itself seems to work just fine now

**Comment by tibbi** [comment]
should be good, thanks

**Pull request merged**